File: Algorithms/MST Prims/Prims.cpp
typedef pair<int, int> ii;

void primsMST(vector<ii> adj[], int V)
{
    int src = 0;
    bool MSTset[V];
    int weights[V], parents[V];
    priority_queue<ii, vector<ii>, greater<ii>> pq;

    fill(MSTset, MSTset + V, false);
    fill(weights, weights + V, INT_MAX);
    fill(parents, parents + V, -1);

    weights[src] = 0;
    pq.push(make_pair(weights[src], src));

    while (!pq.empty())
    {
        int u = pq.top().second;
        pq.pop();

        if (MSTset[u])
            continue;

        MSTset[u] = true;

        for (ii edge : adj[u])
        {
            int v = edge.first;
            int w = edge.second;
            if (!MSTset[v] && weights[v] > w)
            {
                pq.push(make_pair(w, v));
                weights[v] = w;
                parents[v] = u;
            }
        }
    }
}

File: Algorithms/MST Prims/Prims.py
from queue import PriorityQueue


def primsMST(adj):
    V = len(adj)

    src = 0
    MSTset = [False] * V
    weights = [float("inf")] * V
    parents = [-1] * V
    pq = PriorityQueue(V)

    weights[src] = 0
    pq.put((weights[src], src))

    while not pq.empty():
        _, u = pq.get()
        if MSTset[u]:
            continue

        MSTset[u] = True
        for v, w in adj[u]:
            if not MSTset[v] and weights[v] > w:
                pq.put((w, v))
                weights[v] = w
                parents[v] = u
